util/linuxfw/iptable_runner.go

```diff
@@ -0,0 +1,506 @@
+// local subnets.
+func (i *iptablesRunner) AddSNATRule() error {
+
+	args := []string{"-m", "mark", "--mark", TailscaleSubnetRouteMark + "/" + TailscaleFwmarkMask, "-j", "MASQUERADE"}
+	if err := i.ipt4.Append("nat", "ts-postrouting", args...); err != nil {
+		return fmt.Errorf("adding %v in v4/nat/ts-postrouting: %w", args, err)
+	}
+	if i.v6NATAvailable {
+		if err := i.ipt6.Append("nat", "ts-postrouting", args...); err != nil {
+			return fmt.Errorf("adding %v in v6/nat/ts-postrouting: %w", args, err)
+		}
+	}
+	return nil
+}
+
+// delSNATRule removes the netfilter rule to SNAT traffic destined for
+// local subnets. Fails if the rule does not exist.
+func (i *iptablesRunner) DelSNATRule() error {
+
+	args := []string{"-m", "mark", "--mark", TailscaleSubnetRouteMark + "/" + TailscaleFwmarkMask, "-j", "MASQUERADE"}
+	if err := i.ipt4.Delete("nat", "ts-postrouting", args...); err != nil {
+		return fmt.Errorf("deleting %v in v4/nat/ts-postrouting: %w", args, err)
+	}
+	if i.v6NATAvailable {
+		if err := i.ipt6.Delete("nat", "ts-postrouting", args...); err != nil {
+			return fmt.Errorf("deleting %v in v6/nat/ts-postrouting: %w", args, err)
+		}
+	}
+	return nil
+}
+
+// Cleans the iptables rules created by Tailscale.
+func IptablesCleanUp(logf logger.Logf) {
+	// Clear iptables rules
+	err := clearRules(iptables.ProtocolIPv4, logf)
+	if err != nil {
+		logf("ERROR: clear iptables: %w", err)
+	}
+
+	// Clear ip6tables rules
+	err = clearRules(iptables.ProtocolIPv6, logf)
+	if err != nil {
+		// If the system doesn't support ipv6, we'll just log the error and continue
+		logf("ERROR: clear ip6tables: %w", err)
+	}
+
+}
+
+// clearRules clears all the iptables rules created by Tailscale for the given protocol.
+func clearRules(proto iptables.Protocol, logf logger.Logf) error {
+	ipt, err := iptables.NewWithProtocol(proto)
+	if err != nil {
+		return err
+	}
+
+	// Delete all hooks to our chains
+	delHook := func(ipt iptables.IPTables, table, chain string) error {
+		tsChain := tsChain(chain)
+		args := []string{"-j", tsChain}
+		if err := ipt.Delete(table, chain, args...); err != nil {
+			return nil
+		}
+		return nil
+	}
+
+	if err := delHook(*ipt, "filter", "INPUT"); err != nil {
+		logf("ERROR: delete hook: %w", err)
+	}
+	if err := delHook(*ipt, "filter", "FORWARD"); err != nil {
+		logf("ERROR: delete hook: %w", err)
+	}
+	if err := delHook(*ipt, "nat", "POSTROUTING"); err != nil {
+		logf("ERROR: delete hook: %w", err)
+	}
+
+	// Flush and delete all ts-chains
+	delChain := func(ipt iptables.IPTables, table, chain string) error {
+		if err := ipt.ClearChain(table, chain); err != nil {
+			if isErrChainNotExist(err) {
+				// nonexistent chain. nothing to do.
+				return nil
+			}
+			return fmt.Errorf("flushing %s/%s: %w", table, chain, err)
+		}
+		if err := ipt.DeleteChain(table, chain); err != nil {
+			return fmt.Errorf("deleting %s/%s: %v", table, chain, err)
+		}
+		return nil
+	}
```

Comment:
I would move the two functions from this function into their own functions instead.

Reply:
I think this one would probably make sense since they are actually reused and can have meaningful names like delChain/clearChain.